### PR TITLE
Combine InteractiveMarkerViewer classes.

### LINKED ISSUE #519: Clean up rviz namespace
1. Documentation required.
2. Unnecessary code duplication. Example: WorldInteractiveMarkerViewer copies almost everything from the base class. :-1:  :disappointed:

### LINKED ISSUE #520: (World)InteractiveMarkerViewer
@brianhou 

PR #242 introduced `WorldInteractiveMarkerViewer`(WIMV) that derives from `InteractiveMarkerViewer`(IMV). The only difference is that the `update()` function automatically creates markers for the skeletons in the world to visualize them. This led to quite a bit of code duplication. 

1. Let the two classes exist. Perform a little bit of cleanup for WIMV to reduce code duplication.
2. Take a world in IMV defaulted to `nullptr`. If world exists, add the skeletons in it to the viewer, otherwise update existing markers (TSRs and other shape/frame markers).
3. Port all the base class functions to WIMV and delete IMV completely.

I would prefer 3 > 2 > 1.

### PR DESCRIPTION
**Pull request by aditya-vk**
Simplifies the code in InteractiveMarkerViewer. Refer to #520 for details (followed suggestion 2).

Resolves #519. Resolves #520.

***

**Before creating a pull request**

- [x] Document new methods and classes
- [x] Format code with `make format`

**Before merging a pull request**

- [x] Set version target by selecting a milestone on the right side
- [x] Summarize this change in `CHANGELOG.md`
- [ ] Add unit test(s) for this change

### CODE REVIEW

File: src/rviz/InteractiveMarkerViewer.cpp [hunk 1]
```diff
@@ -168,34 +164,52 @@ void InteractiveMarkerViewer::update()
...
+      if (mWorld->hasSkeleton(it->first))
+      {
+        ++it;
+      }
+      // Skeleton does not exist. Delete our existing ShapeFrameMarker.
+      else
+      {
+        it = mSkeletonMarkers.erase(it);
+      }
     }
```
Comment (Kelym) [inline src/rviz/InteractiveMarkerViewer.cpp]: To confirm: we don't need to update `mSkeletonMarkers`even if we don't have `mWorld`?

File: include/aikido/rviz/InteractiveMarkerViewer.hpp [hunk 2]
```diff
@@ -24,8 +25,14 @@ AIKIDO_DECLARE_POINTERS(InteractiveMarkerViewer)
 class InteractiveMarkerViewer
 {
 public:
+  /// Create an InteractiveMarkerViewer that reflects skeletons in a World.
```
Comment (Kelym) [inline include/aikido/rviz/InteractiveMarkerViewer.hpp]:
```suggestion
  /// Creates an InteractiveMarkerViewer that reflects skeletons in a World.
```

File: include/aikido/rviz/InteractiveMarkerViewer.hpp [hunk 3]
```diff
@@ -24,8 +25,14 @@ AIKIDO_DECLARE_POINTERS(InteractiveMarkerViewer)
...
+  /// \param[in] topicNamespace ROS topic to publish marker updates to
+  /// \param[in] frameId Base frame name
+  /// \param[in] env World to update viewer with
```
Comment (Kelym) [inline include/aikido/rviz/InteractiveMarkerViewer.hpp]:
```suggestion
  /// \param[in] env World to update the viewer with
```

File: include/aikido/rviz/InteractiveMarkerViewer.hpp [hunk 4]
```diff
@@ -34,22 +41,29 @@ class InteractiveMarkerViewer
...
-  FrameMarkerPtr addFrame(
+  /// Visualizes a Skeleton.
+  /// \param skeleton Skeleton to add to the viewer
+  /// \return Skeleton marker added to the viewer.
+  SkeletonMarkerPtr addSkeletonMarker(
+      const dart::dynamics::SkeletonPtr& skeleton);
+
+  /// Visualizes a Frame as a cylinder.
+  /// \param[in] frame Target DART frame.
+  /// \param[in] length Length of the cylinderical frame (along z-axis).
```
Comment (Kelym) [inline include/aikido/rviz/InteractiveMarkerViewer.hpp]:
```suggestion
  /// \param[in] length Length of the cylindrical frame (along z-axis).
```

File: include/aikido/rviz/InteractiveMarkerViewer.hpp [hunk 5]
```diff
@@ -34,22 +41,29 @@ class InteractiveMarkerViewer
...
+  /// \param[in] thickness Radius (thickness in x and y axes) of cylinder.
+  /// \param[in] alpha Opacity.
+  /// \return Frame marker added to this viewer.
+  FrameMarkerPtr addFrameMarker(
       dart::dynamics::Frame* frame,
       double length = 0.25,
       double thickness = 0.02,
       double alpha = 1.0);
 
-  SkeletonMarkerPtr CreateSkeletonMarker(
```
Comment (Kelym) [inline include/aikido/rviz/InteractiveMarkerViewer.hpp]: It looks like a handy helper for creating `SkeletonMarker`. You can use it in `addSkeleton` and `update`. Why bother removing it?

File: include/aikido/rviz/InteractiveMarkerViewer.hpp [hunk 6]
```diff
@@ -77,28 +91,51 @@ class InteractiveMarkerViewer
       double thickness = 0.01,
       std::size_t numLineSegments = 16u);
 
+  /// Set viewer auto-updating.
```
Comment (Kelym) [inline include/aikido/rviz/InteractiveMarkerViewer.hpp]:
```suggestion
  /// Sets viewer auto-updating to on (true) or off.
```

File: include/aikido/rviz/InteractiveMarkerViewer.hpp [hunk 7]
```diff
@@ -34,22 +41,29 @@ class InteractiveMarkerViewer
 
   interactive_markers::InteractiveMarkerServer& marker_server();
 
-  SkeletonMarkerPtr addSkeleton(const dart::dynamics::SkeletonPtr& skeleton);
-
-  FrameMarkerPtr addFrame(
+  /// Visualizes a Skeleton.
+  /// \param skeleton Skeleton to add to the viewer
+  /// \return Skeleton marker added to the viewer.
+  SkeletonMarkerPtr addSkeletonMarker(
```
Comment (Kelym) [inline include/aikido/rviz/InteractiveMarkerViewer.hpp]: So if the Viewer is initialized without `mWorld`, can one still call this function? If not, can we update the docstring and set a check inside the function.

File: include/aikido/rviz/InteractiveMarkerViewer.hpp [hunk 8]
```diff
@@ -77,28 +91,51 @@ class InteractiveMarkerViewer
       double thickness = 0.01,
       std::size_t numLineSegments = 16u);
 
+  /// Set viewer auto-updating.
+  /// \param[in] flag Whether to auto-update the viewer.
   void setAutoUpdate(bool flag);
+
+  /// Update viewer with Skeletons from the World and existing markers.
```
Comment (Kelym) [inline include/aikido/rviz/InteractiveMarkerViewer.hpp]:
```suggestion
  /// Updates viewer with Skeletons from the World and existing markers.
```

File: include/aikido/rviz/InteractiveMarkerViewer.hpp [hunk 9]
```diff
@@ -77,28 +91,51 @@ class InteractiveMarkerViewer
...
 
 protected:
+  /// Thread target for auto-updating the viewer.
   void autoUpdate();
 
+  /// Interactive Marker Server.
   interactive_markers::InteractiveMarkerServer mMarkerServer;
-  std::set<SkeletonMarkerPtr> mSkeletonMarkers;
+
+  /// Mapping of Skeletons to SkeletonMarkers
```
Comment (Kelym) [inline include/aikido/rviz/InteractiveMarkerViewer.hpp]:
```suggestion
  /// Map of Skeletons to SkeletonMarkers
```

File: include/aikido/rviz/InteractiveMarkerViewer.hpp (hunk 5 above)
Comment (aditya-vk) [reply include/aikido/rviz/InteractiveMarkerViewer.hpp]: I preferred consistency (across skeletons, frames etc.) over code duplication. We could also have `createFrameMarker`, `createTSRMarker` etc. I just removed it instead.

File: include/aikido/rviz/InteractiveMarkerViewer.hpp (hunk 7 above)
Comment (aditya-vk) [reply include/aikido/rviz/InteractiveMarkerViewer.hpp]: Yes.

**Comment by aditya-vk** [comment]
@Kelym @brianhou Addressed the comments. Added a docstring that `mWorld` can be a `nullptr` if we are just adding markers, and otherwise the skeletons in the world shall also get updated. Could you review?

File: src/rviz/InteractiveMarkerViewer.cpp [hunk 10]
```diff
@@ -168,34 +164,60 @@ void InteractiveMarkerViewer::update()
...
+    else if (it->first)
     {
       std::unique_lock<std::mutex> skeleton_lock(
-          skeleton->getMutex(), std::try_to_lock);
+          it->first->getMutex(), std::try_to_lock);
       if (skeleton_lock.owns_lock())
-        (*it)->update();
-
-      ++it;
+        it->second->update();
```
Comment (Kelym) [inline src/rviz/InteractiveMarkerViewer.cpp]:
```suggestion
        it->second->update();
        ++it;
```

**Comment by aditya-vk** [comment]
@Kelym I updated the formatting and the ChangeLog. Should be ready for merging into master?

File: CHANGELOG.md [hunk 11]
```diff
@@ -82,6 +82,7 @@
 
   * Fixed bug of not joining Viewer threads when stopping auto-update: [#463](https://github.com/personalrobotics/aikido/pull/463)
   * Fixed bug of not passing full file path to RViz when MeshShape is used: [#518](https://github.com/personalrobotics/aikido/pull/518)
+  * Merge WorldInteractiveMarkerViewer and InteractiveMarkerViewer classes, removing the former: [#520](https://github.com/personalrobotics/aikido/pull/520)
```
Comment (brianhou) [inline CHANGELOG.md]:
```suggestion
  * Merged WorldInteractiveMarkerViewer into InteractiveMarkerViewer, removing the former: [#520](https://github.com/personalrobotics/aikido/pull/520)
```

**Pull request merged**